### ISSUE TITLE
react-hooks 5.0.1 vs 5.0.0

### ISSUE THREAD

**Issue by chulander** [styled-components/styled-components#3015]
OS: macOS 10.15.3
CPU: (4) x64 Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
Memory: 82.75 MB / 8.00 GB
Shell: 5.7.1 - /bin/zsh
Binaries:
Node: 11.1.0 - ~/.nvm/versions/node/v11.1.0/bin/node
Yarn: 1.19.1 - /usr/local/bin/yarn
npm: 6.4.1 - ~/.nvm/versions/node/v11.1.0/bin/npm
npmPackages:
babel-plugin-styled-components: ^1.10.6 => 1.10.6
styled-components: 5.0.0 => 5.0.1
Reproduction
https://codesandbox.io/s/react-hooks-styled-components-500rc-3-9dv0p

Steps to reproduce
Change the dependencies from 5.0.0rc3 or 5.0.0 to **5.0.1** and look at the react-hook warnings about calling not calling useMemo/useCallback other react hooks inside each other.

wondering what changed between in version 5.0.1> I've been using the beta versions since last summer and have been updating to the latest beta to 5.0.0rc-3 to stable version 5.0.0. 

Note, I could be using react-hooks incorrectly here but I thought calling useMemo would maintain the same reference during re-renders and not trigger any issues of hooks calling out of order.

thanks!

**Comment by kitten** [comment]
Hiya 👋 we’ve introduced this error message explicitly, since we became aware of the fact that some people weren’t using the API as intended / as directed by the docs, which may break things later down the road. Let me elaborate by breaking down the error 🙂

> The component styled.button with the id of "sc-AxirZ" has been created dynamically.

When you call `styled.button` or any other styled function you are creating a new StyledComponent. These are special components in that we insert styles for them and hence they become “globally unique” so to speak.

> You may see this warning because you've called styled inside another component.

So to summarise this and the last statement, this means that if you create them dynamically, so inside the render function of another component, you’re creating the StyledComponent over and over again, which causes a memory leak. Calling it inside a hook, even `useMemo` is also not allowed as you’re still recreating the StyledComponent.

> To resolve this only create new StyledComponents outside of any render method and function component.

As this says, create your StyledComponent like any other component: not dynamically. 🙌

**Comment by chulander** [comment]
Hi,

Thank you for looking into this, I updated the same to track the reference in memory and useMemo returns the same object in memory meaning that it was not created multiple times.

https://codesandbox.io/s/react-hooks-styled-components-500rc-3-9dv0p

```jsx
import React, { useMemo } from "react";
import { useStyles } from "./useStyles";
import styled from "styled-components";
let trackButtonReference;
let trackStylesReference;
export function Button(props) {
  const styles = useStyles(props);
  if (typeof trackStylesReference === "undefined") {
    trackStylesReference = styles;
    console.log("initial styles");
  } else if (trackStylesReference !== styles) {
    console.log("different styles");
    trackStylesReference = styles;
  } else if (trackStylesReference === styles) {
    console.log("same styles");
  }
  const Something = useMemo(
    () =>
      styled.button`
        ${styles};
      `,
    [styles]
  );

  const Button = <Something {...props} />;
  if (typeof trackButtonReference === "undefined") {
    trackButtonReference = Something;
    console.log("assigning initial button");
  } else if (trackButtonReference === Something) {
    console.log("same button");
  } else if (trackButtonReference !== Something) {
    trackButtonReference = Something;
    console.log("different button");
  }
  return Button;
}

```

**Comment by chulander** [comment]
I've isolated the button using v5.0.0 and during re-rendering, the internal StyledComponentImpl references / doesn't change. Does that mean that the component did not generate not "dynamically" generate a new button but instead used the same button as the previous render?

![image](https://user-images.githubusercontent.com/4703475/74363021-b75dce00-4d97-11ea-923a-e59504ec718a.png)

**Comment by kitten** [comment]
The reist ruction here is artificial. While this may have worked in prior versions it will always be an anti-pattern and will lead to memory leaks.

You can’t circumvent our error message here because it’s put in place to exactly catch this anti-pattern. It doesn’t quite matter whether it’s technically only called once, because it’s unnecessary to create StyledComponents inside other components.

So in conclusion, you’ll always want to write:

```js
const StyledComp = styles.button``;

const YourComp = () => {
  // never create StyledAcomponents inside other components
  return <StyledComp />
};
```

**Comment by chulander** [comment]
thanks for the clarification.

**Comment by centigrade-julian-lang** [comment]
> So to summarise this and the last statement, this means that if you create them dynamically, so inside the render function of another component, you’re creating the StyledComponent over and over again, which causes a memory leak. Calling it inside a hook, even `useMemo` is also not allowed as you’re still recreating the StyledComponent.

Hi, sorry if there is a resource describing this already, but I do not understand how StyledComponents is intended to be used, if it should not be used within a component's render function?

Maybe I'm confused because I am new to both, react and styled components, but as far as I understand, literally anything is somehow within a render function - the render function of the app-root, isn't it? Sorry, if I am horribly wrong, I am just getting started with all this.

Thanks in advance for your feedback!

**Comment by huanlirui** [comment]
> iya 👋我们已经明确引入了此错误消息，因为我们意识到有些人没有按照文档的意图/按照文档的指示使用API​​，这可能会在以后造成麻烦。让我详细解释一下错误🙂
> 
> > ID为“ sc-AxirZ”的组件style.button已动态创建。
> 
> 当您调用`styled.button`或其他任何样式化的函数时，您将创建一个新的StyledComponent。这些是特殊的组件，因为我们为其插入样式，因此它们可以说是“全球唯一”。
> 
> > 您可能会看到此警告，因为您在另一个组件中调用了样式。
> 
> 因此，总结一下这和最后一个语句，这意味着如果动态创建它们，那么在另一个组件的渲染函数中，您将一遍又一遍地创建StyledComponent，这会导致内存泄漏。在挂钩内调用它，甚至`useMemo`也不允许，因为您仍在重新创建StyledComponent。
> 
> > 若要解决此问题，请仅在任何渲染方法和函数组件之外创建新的StyledComponents。
> 
> 如此说来，请像创建其他任何组件一样创建StyledComponent：不要动态创建。 🙌

说得太对了！
